File: output/cpp/tests/34_doc_history_hsm.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>


#include <iostream>
#include <string>
#include <cassert>
#include <vector>
#include <algorithm>

// Documentation Example: HSM with History (History203)

class HistoryHSMFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    HistoryHSMFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class HistoryHSMFrameContext {
public:
    HistoryHSMFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    HistoryHSMFrameContext(HistoryHSMFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class HistoryHSMCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<HistoryHSMFrameEvent> forward_event;
    std::unique_ptr<HistoryHSMCompartment> parent_compartment;

    explicit HistoryHSMCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<HistoryHSMCompartment> clone() const {
        auto c = std::make_unique<HistoryHSMCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class HistoryHSM {
private:
    std::vector<std::unique_ptr<HistoryHSMCompartment>> _state_stack;
    std::unique_ptr<HistoryHSMCompartment> __compartment;
    std::unique_ptr<HistoryHSMCompartment> __next_compartment;
    std::vector<HistoryHSMFrameContext> _context_stack;

    void __kernel(HistoryHSMFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            HistoryHSMFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                HistoryHSMFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    HistoryHSMFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(HistoryHSMFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Waiting") {
            _state_Waiting(__e);
        } else if (state_name == "A") {
            _state_A(__e);
        } else if (state_name == "B") {
            _state_B(__e);
        } else if (state_name == "AB") {
            _state_AB(__e);
        } else if (state_name == "C") {
            _state_C(__e);
        }
    }

    void __transition(std::unique_ptr<HistoryHSMCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_A(HistoryHSMFrameEvent& __e) {
        if (__e._message == "$>") {
            this->log_msg("In $A");
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("A"));
            return;;
        } else if (__e._message == "gotoB") {
            this->log_msg("gotoB");
            auto __new_compartment = std::make_unique<HistoryHSMCompartment>("B");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else {
            _state_AB(__e);
        }
    }

    void _state_Waiting(HistoryHSMFrameEvent& __e) {
        if (__e._message == "$>") {
            this->log_msg("In $Waiting");
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("Waiting"));
            return;;
        } else if (__e._message == "gotoA") {
            this->log_msg("gotoA");
            auto __new_compartment = std::make_unique<HistoryHSMCompartment>("A");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else if (__e._message == "gotoB") {
            this->log_msg("gotoB");
            auto __new_compartment = std::make_unique<HistoryHSMCompartment>("B");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void _state_B(HistoryHSMFrameEvent& __e) {
        if (__e._message == "$>") {
            this->log_msg("In $B");
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("B"));
            return;;
        } else if (__e._message == "gotoA") {
            this->log_msg("gotoA");
            auto __new_compartment = std::make_unique<HistoryHSMCompartment>("A");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else {
            _state_AB(__e);
        }
    }

    void _state_C(HistoryHSMFrameEvent& __e) {
        if (__e._message == "$>") {
            this->log_msg("In $C");
        } else if (__e._message == "get_log") {
            _context_stack.back()._return = std::any(event_log);
            return;;
        } else if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("C"));
            return;;
        } else if (__e._message == "goBack") {
            this->log_msg("goBack");
            auto __popped = std::move(_state_stack.back()); _state_stack.pop_back();
            __transition(std::move(__popped));
            return;
        }
    }

    void _state_AB(HistoryHSMFrameEvent& __e) {
        if (__e._message == "gotoC") {
            this->log_msg("gotoC in $AB");
            _state_stack.push_back(__compartment->clone());
            auto __new_compartment = std::make_unique<HistoryHSMCompartment>("C");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void log_msg(std::string msg) {
                    event_log.push_back(msg);
    }

public:
    std::vector<std::string> event_log = {};

    HistoryHSM() {
        __compartment = std::make_unique<HistoryHSMCompartment>("Waiting");
        HistoryHSMFrameEvent __frame_event("$>");
        HistoryHSMFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void gotoA() {
        HistoryHSMFrameEvent __e("gotoA");
        HistoryHSMFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void gotoB() {
        HistoryHSMFrameEvent __e("gotoB");
        HistoryHSMFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void gotoC() {
        HistoryHSMFrameEvent __e("gotoC");
        HistoryHSMFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void goBack() {
        HistoryHSMFrameEvent __e("goBack");
        HistoryHSMFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::string get_state() {
        HistoryHSMFrameEvent __e("get_state");
        HistoryHSMFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::vector<std::string> get_log() {
        HistoryHSMFrameEvent __e("get_log");
        HistoryHSMFrameContext __ctx(std::move(__e), std::any(std::vector<std::string>()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::vector<std::string>>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 34: Doc History HSM ===\n");
    HistoryHSM h;

    if (h.get_state() != "Waiting") {
        printf("FAIL: Expected 'Waiting', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.gotoA();
    if (h.get_state() != "A") {
        printf("FAIL: Expected 'A', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.gotoC();
    if (h.get_state() != "C") {
        printf("FAIL: Expected 'C', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.goBack();
    if (h.get_state() != "A") {
        printf("FAIL: Expected 'A' after goBack, got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.gotoB();
    if (h.get_state() != "B") {
        printf("FAIL: Expected 'B', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.gotoC();
    if (h.get_state() != "C") {
        printf("FAIL: Expected 'C', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.goBack();
    if (h.get_state() != "B") {
        printf("FAIL: Expected 'B' after goBack, got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    printf("PASS: HSM with history works correctly\n");
    return 0;
}

File: output/go/tests/34_doc_history_hsm.go

package main

import "fmt"
import "os"

type HistoryHSMFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type HistoryHSMFrameContext struct {
    _event  HistoryHSMFrameEvent
    _return any
    _data   map[string]any
}

type HistoryHSMCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *HistoryHSMFrameEvent
    parentCompartment *HistoryHSMCompartment
}

func newHistoryHSMCompartment(state string) *HistoryHSMCompartment {
    return &HistoryHSMCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *HistoryHSMCompartment) copy() *HistoryHSMCompartment {
    nc := &HistoryHSMCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type HistoryHSM struct {
    _state_stack []*HistoryHSMCompartment
    __compartment *HistoryHSMCompartment
    __next_compartment *HistoryHSMCompartment
    _context_stack []HistoryHSMFrameContext
    log []string
}

func NewHistoryHSM() *HistoryHSM {
    s := &HistoryHSM{}
    s._state_stack = make([]*HistoryHSMCompartment, 0)
    s._context_stack = make([]HistoryHSMFrameContext, 0)
    s.__compartment = newHistoryHSMCompartment("Waiting")
    s.__next_compartment = nil
    __frame_event := HistoryHSMFrameEvent{_message: "$>", _parameters: nil}
    __ctx := HistoryHSMFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *HistoryHSM) __kernel(__e *HistoryHSMFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &HistoryHSMFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &HistoryHSMFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &HistoryHSMFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *HistoryHSM) __router(__e *HistoryHSMFrameEvent) {
    switch s.__compartment.state {
    case "Waiting":
        s._state_Waiting(__e)
    case "A":
        s._state_A(__e)
    case "B":
        s._state_B(__e)
    case "AB":
        s._state_AB(__e)
    case "C":
        s._state_C(__e)
    }
}

func (s *HistoryHSM) __transition(next *HistoryHSMCompartment) {
    s.__next_compartment = next
}

func (s *HistoryHSM) GotoA() {
    __e := HistoryHSMFrameEvent{_message: "GotoA"}
    __ctx := HistoryHSMFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryHSM) GotoB() {
    __e := HistoryHSMFrameEvent{_message: "GotoB"}
    __ctx := HistoryHSMFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryHSM) GotoC() {
    __e := HistoryHSMFrameEvent{_message: "GotoC"}
    __ctx := HistoryHSMFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryHSM) GoBack() {
    __e := HistoryHSMFrameEvent{_message: "GoBack"}
    __ctx := HistoryHSMFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryHSM) GetState() string {
    __e := HistoryHSMFrameEvent{_message: "GetState"}
    __ctx := HistoryHSMFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HistoryHSM) GetLog() []string {
    __e := HistoryHSMFrameEvent{_message: "GetLog"}
    __ctx := HistoryHSMFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result []string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.([]string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HistoryHSM) _state_B(__e *HistoryHSMFrameEvent) {
    if __e._message == "$>" {
        s.__log_msg("In $B")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "B"
        return
    } else if __e._message == "GotoA" {
        s.__log_msg("gotoA")
        __compartment := newHistoryHSMCompartment("A")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else {
        s._state_AB(__e)
    }
}

func (s *HistoryHSM) _state_AB(__e *HistoryHSMFrameEvent) {
    if __e._message == "GotoC" {
        s.__log_msg("gotoC in $AB")
        s._state_stack = append(s._state_stack, s.__compartment.copy())
        __compartment := newHistoryHSMCompartment("C")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *HistoryHSM) _state_C(__e *HistoryHSMFrameEvent) {
    if __e._message == "$>" {
        s.__log_msg("In $C")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "C"
        return
    } else if __e._message == "GoBack" {
        s.__log_msg("goBack")
        __popped := s._state_stack[len(s._state_stack)-1]
        s._state_stack = s._state_stack[:len(s._state_stack)-1]
        s.__transition(__popped)
        return
    }
}

func (s *HistoryHSM) _state_Waiting(__e *HistoryHSMFrameEvent) {
    if __e._message == "$>" {
        s.__log_msg("In $Waiting")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Waiting"
        return
    } else if __e._message == "GotoA" {
        s.__log_msg("gotoA")
        __compartment := newHistoryHSMCompartment("A")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "GotoB" {
        s.__log_msg("gotoB")
        __compartment := newHistoryHSMCompartment("B")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *HistoryHSM) _state_A(__e *HistoryHSMFrameEvent) {
    if __e._message == "$>" {
        s.__log_msg("In $A")
    } else if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "A"
        return
    } else if __e._message == "GotoB" {
        s.__log_msg("gotoB")
        __compartment := newHistoryHSMCompartment("B")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else {
        s._state_AB(__e)
    }
}

func (s *HistoryHSM) __log_msg(msg string) {
                s.log = append(s.log, msg)
}

func main() {
	fmt.Println("=== Test 34: Doc History HSM ===")
	h := NewHistoryHSM()

	// Start in Waiting
	if h.GetState() != "Waiting" {
		fmt.Printf("FAIL: Expected 'Waiting', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go to A
	h.GotoA()
	if h.GetState() != "A" {
		fmt.Printf("FAIL: Expected 'A', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go to C (using inherited GotoC from $AB)
	h.GotoC()
	if h.GetState() != "C" {
		fmt.Printf("FAIL: Expected 'C', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go back (should pop to A)
	h.GoBack()
	if h.GetState() != "A" {
		fmt.Printf("FAIL: Expected 'A' after GoBack, got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go to B
	h.GotoB()
	if h.GetState() != "B" {
		fmt.Printf("FAIL: Expected 'B', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go to C (again using inherited GotoC)
	h.GotoC()
	if h.GetState() != "C" {
		fmt.Printf("FAIL: Expected 'C', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go back (should pop to B)
	h.GoBack()
	if h.GetState() != "B" {
		fmt.Printf("FAIL: Expected 'B' after GoBack, got '%s'\n", h.GetState())
		os.Exit(1)
	}

	fmt.Printf("Log: %v\n", h.GetLog())
	fmt.Println("PASS: HSM with history works correctly")
}
